### PR TITLE
feat: add Game model

### PR DESCRIPTION
**Pull request by qoosuperman**
## Pull Request Description

Create draft Game model and some other relative models

### CODE REVIEW

File: core/src/builders/componentsetbuilder.ts [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    const colorsMap: MonetaryDenominations = MonetaryColors.reduce((map, color) => {
+      map.set(color, tokensCount);
+      return map;
+    }, new Map<string, number>());
+    this.tokens = new MonetaryValue(colorsMap);
+    return this;
+  }
+
+  build(): ComponentSet {
+    return new ComponentSet(this.cards, this.nobles, this.tokens);
```
Comment (qoosuperman) [inline core/src/builders/componentsetbuilder.ts:43]: I think there would be `create()` function if the model need to exist in persistence layer

File: core/src/componentset.ts [hunk 2]
```diff
@@ -1,15 +1,19 @@
 import { Card } from "./card";
+import { MonetaryValue } from "./monetaryvalue";
 import { Noble } from "./noble";
-import { Token } from "./token";
 
 export class ComponentSet {
   cards: Card[];
   nobles: Noble[];
-  tokens: Token[];
+  _tokens: MonetaryValue;
```
Comment (qoosuperman) [inline core/src/componentset.ts]: 我在想是不是改寫成這樣比較可以避免內容被改掉？

File: core/src/monetaryvalue.ts [hunk 3]
```diff
@@ -5,6 +5,8 @@
  */
 export type MonetaryDenominations = InstanceType<typeof Map<string, number>>;
 
+export const MonetaryColors = ["red", "blue", "green", "white", "black", "gold"];
+
```
Comment (qwwqe) [inline core/src/monetaryvalue.ts]: 這個寫成 enum 會不會比較合適？

File: core/src/deck.ts [hunk 4]
```diff
@@ -20,4 +28,12 @@ export class Deck {
   draw() {
     return this._cards.pop() as Card;
   }
+
+  first(): Card {
+    return this._cards[0];
+  }
```
Comment (qwwqe) [inline core/src/deck.ts]: 如果將 `first` 視為牌堆的一種屬性的話，或許可以跟 `level` 和 `size` 一樣做成 getter？

然後想到這邊和 `draw()` 感覺應該回傳 `Card | undefined`，畢竟有機會牌堆是空的，像是四人局狂撿便宜卡曾經遇過第一級卡牌全被拿走

File: core/src/componentset.ts (hunk 2 above)
Comment (qwwqe) [reply core/src/componentset.ts]: 覺得可以～ 但包成 getter 的話，好像屬性本身可以直接下 `private`

File: core/src/game.ts [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+        case 2:
+          deck2.add(card);
+          break;
+        case 3:
+          deck3.add(card);
+          break;
+      }
+    });
+    this.decks = [deck1, deck2, deck3];
+    this.faceUpCards = [deck1.first(), deck2.first(), deck3.first()];
```
Comment (qwwqe) [inline core/src/game.ts]: Shouldn't this be `deck.draw()` as opposed to `deck.first()`, and four calls each?

Also maybe we should add an optional argument to `Deck.draw()` so you could call `deck.draw(4)`...

```
this.faceUpCards = [deck1.draw(4), deck2.draw(4), deck3.draw(4)]
```

File: core/src/game.ts [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+          break;
+        case 2:
+          deck2.add(card);
+          break;
+        case 3:
+          deck3.add(card);
+          break;
+      }
+    });
+    this.decks = [deck1, deck2, deck3];
```
Comment (qwwqe) [inline core/src/game.ts]: I wonder if we could avoid fixing the number of decks altogether? Use a Map or something to record each level's cards, then order the cards by level?

```ts
const decksByLevel = new Map<number, Deck>()
event.cards.forEach((card) => {
  if (!decksByLevel.has(card.level)) {
    decksByLevel.set(card.level, new Deck())
  }
  
  decksByLevel.get(card.level).add(card)
}

this.decks = [...decksByLevel.entries()]
             .sort(([levelA, _], [levelB, _]) => levelA - levelB)
             .map(([_, deck]) => deck)
             
this.faceUpCards = this.decks.map((deck) => deck.draw(4))
```

Or something like that?

File: core/src/game.ts [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+      this.events.push(event);
+    });
+  }
+
+  handleCreatedGameEvent(event: GameCreatedEvent) {
+    this.players = new Array(event.playersCount).fill(new Player());
+    this.tokens = event.tokens;
+    event.nobles.forEach((noble) => {
+      this.nobles.push(noble);
+    });
```
Comment (qwwqe) [inline core/src/game.ts]: 好像可以寫成 `this.nobles = [...event.nobles]` 會是一樣的效果？

除非特別要允許 `GameCreatedEvent` 產生之前，`Game` 可以已經有 `nobles` 了

File: core/src/componentset.ts (hunk 2 above)
Comment (qoosuperman) [reply core/src/componentset.ts]: 我一直忘記屬性可以加上 private 這件事情 = =

File: core/src/game.ts (hunk 7 above)
Comment (qoosuperman) [reply core/src/game.ts]: 覺得 js 的 code 對我來說都超不直觀的 XDDDD
如果是 Ruby 的話應該像是: 
```ruby
this.nobles = event.nobles.dup();
```

File: core/src/componentset.ts (hunk 2 above)
Comment (qoosuperman) [reply core/src/componentset.ts]: [fixed](https://github.com/qoosuperman/panoply/pull/9/commits/915484ad6fe87f4628545a9e490b64341e04fa51)

File: core/src/monetaryvalue.ts (hunk 3 above)
Comment (qoosuperman) [reply core/src/monetaryvalue.ts]: [fixed](https://github.com/qoosuperman/panoply/pull/9/commits/e56beca7736a8dbcf8533ebeaed76e8b2fc90792)

File: core/src/deck.ts (hunk 4 above)
Comment (qoosuperman) [reply core/src/deck.ts]: [fixed](https://github.com/qoosuperman/panoply/pull/9/commits/15f3939db55610d9ec1838e84642638eaa8bce9f)

File: core/src/game.ts (hunk 6 above)
Comment (qoosuperman) [reply core/src/game.ts]: [fixed](https://github.com/qoosuperman/panoply/pull/9/commits/da84274fa623eeb7222b7c26cbc03487905f3497)

File: core/src/game.ts (hunk 7 above)
Comment (qoosuperman) [reply core/src/game.ts]: [fixed](https://github.com/qoosuperman/panoply/pull/9/commits/e6e185d106a219f5150949691cbcb41cf0bba3cc)

**Review by qwwqe — APPROVED** [review]

File: core/src/game.ts (hunk 7 above)
Comment (qwwqe) [reply core/src/game.ts]: js 很多很醜的語法
像是遍歷 0 到 N：`new Array(n).fill().forEach(() => ...)`

**Pull request merged**